lib/improver/threshold.py

```diff
@@ -173,6 +188,16 @@ def process(self, input_cube):
         if np.isnan(input_cube.data).any():
             raise ValueError("Error: NaN detected in input cube data")
 
+        # if necessary, convert thresholds and fuzzy bounds into cube units
+        if self.threshold_units is not None:
+            self.thresholds = [self.threshold_units.convert(threshold,
+                                                            input_cube.units)
+                               for threshold in self.thresholds]
+            self.fuzzy_bounds = [tuple([
+                self.threshold_units.convert(threshold, input_cube.units)
+                for threshold in bounds]) for bounds in self.fuzzy_bounds]
+
+        # apply fuzzy thresholding
```

Comment:
I think that this section needs a lot of inline comments to make it clear what each step is - for example:
`
# Generate truth value array.
for threshold, bounds in zip(self.thresholds, self.fuzzy_bounds):
	cube = input_cube.copy()
	# If no fuzzy factor set - then the upper/lower bounds
	# defaults to be equal to the threshold.
	# truth_value=True where data > threshold and False elsewhere.
	if bounds[0] == bounds[1]:
		truth_value = cube.data > threshold
	# If fuzzy factor set - then the upper/lower bounds are created using
	# threshold*+/-fuzzy factor.

	# The "truth_value" array's output values will be between 0 and 1.
	# 0 represents where the data is below the lower bound and 1 where
	# the data is above the upper bound.
	# Therefore a value which is halfway between the bounds will be assigned 
	# a "truth_value" of 0.5.
	else:
		truth_value = np.where(
			cube.data < threshold,
			rescale(cube.data,  # If data below threshold - scale from 0 to 0.5
					data_range=(bounds[0], threshold),
					scale_range=(0., 0.5),
					clip=True),
			rescale(cube.data,  # If data above threshold - scale from 0.5 to 1
					data_range=(threshold, bounds[1]),
					scale_range=(0.5, 1.),
					clip=True),
		)
	truth_value = truth_value.astype(input_cube_dtype)
`

Reply:
But other than that - I'm happy with these changes.
Cheers, Mark